File: backend/app/schemas/announcement.py
"""Pydantic schemas for announcements (Phase 13)."""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

AnnouncementTargetLiteral = Literal["all", "subject", "college"]


class AnnouncementCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    body: str = Field(..., min_length=1, max_length=5000)
    target: AnnouncementTargetLiteral = "all"
    subject_id: uuid.UUID | None = None  # required if target == 'subject'


class AnnouncementUpdate(BaseModel):
    title: str | None = Field(None, max_length=255)
    body: str | None = Field(None, max_length=5000)


class AnnouncementResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    author_id: uuid.UUID
    author_name: str | None = None
    subject_id: uuid.UUID | None = None
    subject_code: str | None = None
    subject_name: str | None = None
    college_id: uuid.UUID | None = None
    title: str
    body: str
    target: AnnouncementTargetLiteral
    created_at: datetime

File: frontend/src/types/announcement.ts
/**
 * Announcement + analytics types — mirror backend Pydantic schemas in
 * backend/app/schemas/announcement.py and analytics.py
 */

export type AnnouncementTarget = 'all' | 'subject' | 'college'

export interface Announcement {
  id: string
  author_id: string
  author_name: string | null
  subject_id: string | null
  subject_code: string | null
  subject_name: string | null
  college_id: string | null
  title: string
  body: string
  target: AnnouncementTarget
  created_at: string
}

export interface AnnouncementCreate {
  title: string
  body: string
  target: AnnouncementTarget
  subject_id?: string | null
}

export interface AnnouncementUpdate {
  title?: string
  body?: string
}

// ── Analytics ──

export type Trend = 'up' | 'down' | 'flat'

export interface AnalyticsHeadline {
  class_average: number | null
  completion_rate: number | null
  most_missed_topic: string | null
  active_students: number
}

export interface WeakTopicRow {
  topic: string
  score_percent: number
  attempts: number
  subject_code: string | null
}

export interface ScoreBucket {
  label: string
  range_start: number
  range_end: number
  count: number
}

export interface StudentPerformanceRow {
  student_id: string
  name: string
  quizzes_taken: number
  avg_score: number | null
  last_attempt_at: string | null
  trend: Trend
  weak_area: string | null
}

export interface ClassAnalytics {
  teacher_id: string
  subject_id: string | null
  headline: AnalyticsHeadline
  weak_topics: WeakTopicRow[]
  score_distribution: ScoreBucket[]
  student_performance: StudentPerformanceRow[]
}
